src/lib.rs

```diff
@@ -77,6 +77,64 @@ use crate::r#impl::splice::make_splice_iterator;
 
 pub use crate::r#impl::{Drain, DrainFilter, IntoIter, Splice};
 
+#[inline]
+#[allow(clippy::cast_ptr_alignment)]
+fn is_default(buf: core::ptr::NonNull<u8>) -> bool {
+    core::ptr::eq(buf.as_ptr(), &DEFAULT_U8)
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+//TODO: Switch to https://doc.rust-lang.org/alloc/collections/enum.TryReserveErrorKind.html once stable
+/// The error type for `try_reserve` methods.
+pub struct TryReserveError {
+}
+
+struct Grow {
+    buf: core::ptr::NonNull<u8>,
+    old_capacity: usize,
+    new_capacity: usize,
+    alignment: usize,
+    len: usize,
+    type_size: usize,
+}
```

Comment:
I don't get why we have a `Grow` struct that also contains a `type_size`.

I'm not sure I understand this refactoring in general.

We should basically just update `fn grow()` to return a `Option<()>` instead of what we're doing now which is invoking `handle_alloc_error`.

To this end, the current usages just need to match on `grow()`'s new return type and invoke `handle_alloc_error` there and then the `try_reserve_*` associated functions can match on the `None` and return the `Result` type.

This should be largely good enough because any other allocation failures (for example, our layout is invalid) would be caused by bugs in the library.

Reply:
This Grow struct here is to just to make sure not to make mistake when passing arguments to `grow` function as I moved it outside of Vec<T> as part of refactoring.

This is ofc your choice if you do not need refactoring to remove unneeded generic code.
As you said it would be simply done by refactoring original `grow` function

Reply:
To me it seems silly to create a struct just to give it an `impl` with a single associated function.

This pattern, imo, fits better for `DropGuard`-like types.